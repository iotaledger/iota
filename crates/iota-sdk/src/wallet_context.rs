// Copyright (c) Mysten Labs, Inc.
// Modifications Copyright (c) 2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::{collections::BTreeSet, path::Path, sync::Arc};

use anyhow::anyhow;
use colored::Colorize;
use iota_config::{Config, PersistedConfig};
use iota_json_rpc_types::{
    IotaObjectData, IotaObjectDataFilter, IotaObjectDataOptions, IotaObjectResponse,
    IotaObjectResponseQuery, IotaTransactionBlockResponse, IotaTransactionBlockResponseOptions,
};
use iota_keys::keystore::AccountKeystore;
use iota_types::{
    base_types::{IotaAddress, ObjectID, ObjectRef},
    gas_coin::GasCoin,
    transaction::{Transaction, TransactionData, TransactionDataAPI},
};
use shared_crypto::intent::Intent;
use tokio::sync::RwLock;
use tracing::warn;

use crate::{iota_client_config::IotaClientConfig, IotaClient};

pub struct WalletContext {
    pub config: PersistedConfig<IotaClientConfig>,
    request_timeout: Option<std::time::Duration>,
    client: Arc<RwLock<Option<IotaClient>>>,
    max_concurrent_requests: Option<u64>,
}

impl WalletContext {
    pub fn new(
        config_path: &Path,
        request_timeout: Option<std::time::Duration>,
        max_concurrent_requests: Option<u64>,
    ) -> Result<Self, anyhow::Error> {
        let config: IotaClientConfig = PersistedConfig::read(config_path).map_err(|err| {
            anyhow!("Cannot open wallet config file at {config_path:?}. Err: {err}",)
        })?;

        let config = config.persisted(config_path);
        let context = Self {
            config,
            request_timeout,
            client: Default::default(),
            max_concurrent_requests,
        };
        Ok(context)
    }

    pub fn get_addresses(&self) -> Vec<IotaAddress> {
        self.config.keystore.addresses()
    }

    pub async fn get_client(&self) -> Result<IotaClient, anyhow::Error> {
        let read = self.client.read().await;

        Ok(if let Some(client) = read.as_ref() {
            client.clone()
        } else {
            drop(read);
            let client = self
                .config
                .get_active_env()?
                .create_rpc_client(self.request_timeout, self.max_concurrent_requests)
                .await?;
            if let Err(e) = client.check_api_version() {
                warn!("{e}");
                eprintln!("{}", format!("[warn] {e}").yellow().bold());
            }
            self.client.write().await.insert(client).clone()
        })
    }

    // TODO: Ger rid of mut
    pub fn active_address(&mut self) -> Result<IotaAddress, anyhow::Error> {
        if self.config.keystore.addresses().is_empty() {
            return Err(anyhow!(
                "No managed addresses. Create new address with `new-address` command."
            ));
        }

        // Ok to unwrap because we checked that config addresses not empty
        // Set it if not exists
        self.config.active_address = Some(
            self.config
                .active_address
                .unwrap_or(*self.config.keystore.addresses().first().unwrap()),
        );

        Ok(self.config.active_address.unwrap())
    }

    /// Get the latest object reference given a object id
    pub async fn get_object_ref(&self, object_id: ObjectID) -> Result<ObjectRef, anyhow::Error> {
        let client = self.get_client().await?;
        Ok(client
            .read_api()
            .get_object_with_options(object_id, IotaObjectDataOptions::new())
            .await?
            .into_object()?
            .object_ref())
    }

    /// Get all the gas objects (and conveniently, gas amounts) for the address
    pub async fn gas_objects(
        &self,
        address: IotaAddress,
    ) -> Result<Vec<(u64, IotaObjectData)>, anyhow::Error> {
        let client = self.get_client().await?;

        let mut objects: Vec<IotaObjectResponse> = Vec::new();
        let mut cursor = None;
        loop {
            let response = client
                .read_api()
                .get_owned_objects(
                    address,
                    Some(IotaObjectResponseQuery::new(
                        Some(IotaObjectDataFilter::StructType(GasCoin::type_())),
                        Some(IotaObjectDataOptions::full_content()),
                    )),
                    cursor,
                    None,
                )
                .await?;

            objects.extend(response.data);

            if response.has_next_page {
                cursor = response.next_cursor;
            } else {
                break;
            }
        }

        // TODO: We should ideally fetch the objects from local cache
        let mut values_objects = Vec::new();

        for object in objects {
            let o = object.data;
            if let Some(o) = o {
                let gas_coin = GasCoin::try_from(&o)?;
                values_objects.push((gas_coin.value(), o.clone()));
            }
        }

        Ok(values_objects)
    }

    pub async fn get_object_owner(&self, id: &ObjectID) -> Result<IotaAddress, anyhow::Error> {
        let client = self.get_client().await?;
        let object = client
            .read_api()
            .get_object_with_options(*id, IotaObjectDataOptions::new().with_owner())
            .await?
            .into_object()?;
        Ok(object
            .owner
            .ok_or_else(|| anyhow!("Owner field is None"))?
            .get_owner_address()?)
    }

    pub async fn try_get_object_owner(
        &self,
        id: &Option<ObjectID>,
    ) -> Result<Option<IotaAddress>, anyhow::Error> {
        if let Some(id) = id {
            Ok(Some(self.get_object_owner(id).await?))
        } else {
            Ok(None)
        }
    }

    /// Find a gas object which fits the budget
    pub async fn gas_for_owner_budget(
        &self,
        address: IotaAddress,
        budget: u64,
        forbidden_gas_objects: BTreeSet<ObjectID>,
    ) -> Result<(u64, IotaObjectData), anyhow::Error> {
        for o in self.gas_objects(address).await.unwrap() {
            if o.0 >= budget && !forbidden_gas_objects.contains(&o.1.object_id) {
                return Ok((o.0, o.1));
            }
        }
        Err(anyhow!(
            "No non-argument gas objects found for this address with value >= budget {budget}. Run iota client gas to check for gas objects."
        ))
    }

    pub async fn get_all_gas_objects_owned_by_address(
        &self,
        address: IotaAddress,
    ) -> anyhow::Result<Vec<ObjectRef>> {
        self.get_gas_objects_owned_by_address(address, None).await
    }

    pub async fn get_gas_objects_owned_by_address(
        &self,
        address: IotaAddress,
        limit: Option<usize>,
    ) -> anyhow::Result<Vec<ObjectRef>> {
        let client = self.get_client().await?;
        let results: Vec<_> = client
            .read_api()
            .get_owned_objects(
                address,
                Some(IotaObjectResponseQuery::new(
                    Some(IotaObjectDataFilter::StructType(GasCoin::type_())),
                    Some(IotaObjectDataOptions::full_content()),
                )),
                None,
                limit,
            )
            .await?
            .data
            .into_iter()
            .filter_map(|r| r.data.map(|o| o.object_ref()))
            .collect();
        Ok(results)
    }

    /// Given an address, return one gas object owned by this address.
    /// The actual implementation just returns the first one returned by the
    /// read api.
    pub async fn get_one_gas_object_owned_by_address(
        &self,
        address: IotaAddress,
    ) -> anyhow::Result<Option<ObjectRef>> {
        Ok(self
            .get_gas_objects_owned_by_address(address, Some(1))
            .await?
            .pop())
    }

    /// Returns one address and all gas objects owned by that address.
    pub async fn get_one_account(&self) -> anyhow::Result<(IotaAddress, Vec<ObjectRef>)> {
        let address = self.get_addresses().pop().unwrap();
        Ok((
            address,
            self.get_all_gas_objects_owned_by_address(address).await?,
        ))
    }

    /// Return a gas object owned by an arbitrary address managed by the wallet.
    pub async fn get_one_gas_object(&self) -> anyhow::Result<Option<(IotaAddress, ObjectRef)>> {
        for address in self.get_addresses() {
            if let Some(gas_object) = self.get_one_gas_object_owned_by_address(address).await? {
                return Ok(Some((address, gas_object)));
            }
        }
        Ok(None)
    }

    /// Returns all the account addresses managed by the wallet and their owned
    /// gas objects.
    pub async fn get_all_accounts_and_gas_objects(
        &self,
    ) -> anyhow::Result<Vec<(IotaAddress, Vec<ObjectRef>)>> {
        let mut result = vec![];
        for address in self.get_addresses() {
            let objects = self
                .gas_objects(address)
                .await?
                .into_iter()
                .map(|(_, o)| o.object_ref())
                .collect();
            result.push((address, objects));
        }
        Ok(result)
    }

    pub async fn get_reference_gas_price(&self) -> Result<u64, anyhow::Error> {
        let client = self.get_client().await?;
        let gas_price = client.governance_api().get_reference_gas_price().await?;
        Ok(gas_price)
    }

    /// Sign a transaction with a key currently managed by the WalletContext
    pub fn sign_transaction(&self, data: &TransactionData) -> Transaction {
        let sig = self
            .config
            .keystore
            .sign_secure(&data.sender(), data, Intent::iota_transaction())
            .unwrap();
        // TODO: To support sponsored transaction, we should also look at the gas owner.
        Transaction::from_data(data.clone(), vec![sig])
    }

    /// Execute a transaction and wait for it to be locally executed on the
    /// fullnode. Also expects the effects status to be
    /// ExecutionStatus::Success.
    pub async fn execute_transaction_must_succeed(
        &self,
        tx: Transaction,
    ) -> IotaTransactionBlockResponse {
        tracing::debug!("Executing transaction: {:?}", tx);
        let response = self.execute_transaction_may_fail(tx).await.unwrap();
        assert!(
            response.status_ok().unwrap(),
            "Transaction failed: {:?}",
            response
        );
        response
    }

    /// Execute a transaction and wait for it to be locally executed on the
    /// fullnode. The transaction execution is not guaranteed to succeed and
    /// may fail. This is usually only needed in non-test environment or the
    /// caller is explicitly testing some failure behavior.
    pub async fn execute_transaction_may_fail(
        &self,
        tx: Transaction,
    ) -> anyhow::Result<IotaTransactionBlockResponse> {
        let client = self.get_client().await?;
        Ok(client
            .quorum_driver_api()
            .execute_transaction_block(
                tx,
                IotaTransactionBlockResponseOptions::new()
                    .with_effects()
                    .with_input()
                    .with_events()
                    .with_object_changes()
                    .with_balance_changes(),
                Some(iota_types::quorum_driver_types::ExecuteTransactionRequestType::WaitForLocalExecution),
            )
            .await?)
    }
}
