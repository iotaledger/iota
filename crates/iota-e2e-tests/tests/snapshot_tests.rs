// Copyright (c) Mysten Labs, Inc.
// Modifications Copyright (c) 2024 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use clap::Parser;
use insta::assert_json_snapshot;
use iota::iota_commands::IotaCommand;
use iota_macros::sim_test;
use iota_sdk::wallet_context::WalletContext;
use test_cluster::TestClusterBuilder;

// special constants for substitution in commands
const ME: &str = "{ME}";

async fn run_one(
    test: Vec<&str>,
    context: &mut WalletContext,
) -> Result<Vec<serde_json::Value>, anyhow::Error> {
    let mut test_output = Vec::new();
    let active_addr = context.active_address()?.to_string();
    for cli_cmd in test {
        let mut cli_cmd_vec = cli_cmd.split(' ').collect::<Vec<&str>>();
        for word in cli_cmd_vec.iter_mut() {
            if *word == ME {
                *word = &active_addr
            }
        }
        test_output.push(serde_json::Value::String(cli_cmd.to_string()));
        let c = IotaCommand::try_parse_from(cli_cmd_vec)?;
        match c {
            IotaCommand::Client { cmd, .. } => {
                if let Some(client_cmd) = cmd {
                    match client_cmd.execute(context).await {
                        Ok(output) => {
                            if let Some(block_response) = output.tx_block_response() {
                                test_output.push(serde_json::to_value(block_response)?);
                            } else if let Some(objects_response) = output.objects_response() {
                                test_output.push(serde_json::to_value(objects_response)?)
                            }
                        }
                        Err(e) => test_output.push(serde_json::Value::String(e.to_string())),
                    }
                }
            }
            IotaCommand::Move {
                package_path: _,
                build_config: _,
                cmd: _,
            } => unimplemented!("Supporting Move publish and upgrade commands"),
            _ => panic!("Command {:?} not supported by RPC snapshot tests", cli_cmd),
        }
    }
    Ok(test_output)
}

#[ignore]
#[sim_test]
async fn basic_read_cmd_snapshot_tests() -> Result<(), anyhow::Error> {
    let mut test_cluster = TestClusterBuilder::new().build().await;
    let context = &mut test_cluster.wallet;

    let cmds = vec![
        "iota client objects {ME}", // valid addr
        "iota client objects 0x0000000000000000000000000000000000000000000000000000000000000000", /* empty addr */
        "iota client object 0x5",       // valid object
        "iota client object 0x5 --bcs", // valid object BCS
        // Simtest object IDs are not stable so these object IDs may or may not exist currently --
        // commenting them out for now.
        // "iota client object 0x3b5121a0603ef7ab4cb57827fceca17db3338ef2cd76126cc1523b681df27cee",
        // // valid object "iota client object
        // 0x3b5121a0603ef7ab4cb57827fceca17db3338ef2cd76126cc1523b681df27cee --bcs", // valid
        // object BCS
        "iota client object 0x0000000000000000000000000000000000000000000000000000000000000000", /* non-existent object */
        "iota client tx-block Duwr9uSk9ZvAndEa8oDHunx345i6oyrp3e78MYHVAbYdv", // valid tx digest
        "iota client tx-block EgMTHQygMi6SRsBqrPHAEKZCNrpShXurCp9rcb9qbSg8",  /* non-existent tx
                                                                              * digest */
    ];
    assert_json_snapshot!(run_one(cmds, context).await?);
    Ok(())
}
