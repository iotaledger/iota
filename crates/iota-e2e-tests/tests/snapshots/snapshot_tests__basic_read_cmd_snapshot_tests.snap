---
source: crates/iota-e2e-tests/tests/snapshot_tests.rs
expression: "run_one(cmds, context).await?"
---
[
  "iota client objects {ME}",
  [
    {
      "data": {
        "objectId": "0x239c0e67c1d0237586af6dec711bd7c6c9496e7ac8b35e3e6df24c27030559da",
        "version": "1",
        "digest": "8nL2C1HMRSjubtD8XeH9beBs3gHz2soBp6cZph8YAA8j",
        "type": "0x2::coin::Coin<0x2::iota::IOTA>",
        "owner": {
          "AddressOwner": "0x00a059cf7241f92b9c716dd5bc2077b32a180a784d907359d007f00cb76970e1"
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "content": {
          "dataType": "moveObject",
          "type": "0x2::coin::Coin<0x2::iota::IOTA>",
          "fields": {
            "balance": "30000000000000000",
            "id": {
              "id": "0x239c0e67c1d0237586af6dec711bd7c6c9496e7ac8b35e3e6df24c27030559da"
            }
          }
        }
      }
    },
    {
      "data": {
        "objectId": "0x4b1561b050cd8dab26de6daf853081c24490a0cdd42b066ceac3e280504e6735",
        "version": "1",
        "digest": "EKKFoQDJJ6pa2GmeNA1yNSzCP76pZX1o77svZTbFxEL7",
        "type": "0x2::coin::Coin<0x2::iota::IOTA>",
        "owner": {
          "AddressOwner": "0x00a059cf7241f92b9c716dd5bc2077b32a180a784d907359d007f00cb76970e1"
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "content": {
          "dataType": "moveObject",
          "type": "0x2::coin::Coin<0x2::iota::IOTA>",
          "fields": {
            "balance": "30000000000000000",
            "id": {
              "id": "0x4b1561b050cd8dab26de6daf853081c24490a0cdd42b066ceac3e280504e6735"
            }
          }
        }
      }
    },
    {
      "data": {
        "objectId": "0xaf5d50b4ba22b3386f32d58522096651967fc9170518d7138942240cccbd759a",
        "version": "1",
        "digest": "CpfG5nBqBarBk9vBV6mKUgEJ5REc99sBmQBrnsn2qE3P",
        "type": "0x2::coin::Coin<0x2::iota::IOTA>",
        "owner": {
          "AddressOwner": "0x00a059cf7241f92b9c716dd5bc2077b32a180a784d907359d007f00cb76970e1"
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "content": {
          "dataType": "moveObject",
          "type": "0x2::coin::Coin<0x2::iota::IOTA>",
          "fields": {
            "balance": "30000000000000000",
            "id": {
              "id": "0xaf5d50b4ba22b3386f32d58522096651967fc9170518d7138942240cccbd759a"
            }
          }
        }
      }
    },
    {
      "data": {
        "objectId": "0xb01af1f557df5884b12d0c8f64bf0b9e7c8a519673742434b0dd4024841de1a5",
        "version": "1",
        "digest": "4m7ojgZzqsHWQDXv7W8ia7jKar2zAuK4gHHXkCyX7UaN",
        "type": "0x2::coin::Coin<0x2::iota::IOTA>",
        "owner": {
          "AddressOwner": "0x00a059cf7241f92b9c716dd5bc2077b32a180a784d907359d007f00cb76970e1"
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "content": {
          "dataType": "moveObject",
          "type": "0x2::coin::Coin<0x2::iota::IOTA>",
          "fields": {
            "balance": "30000000000000000",
            "id": {
              "id": "0xb01af1f557df5884b12d0c8f64bf0b9e7c8a519673742434b0dd4024841de1a5"
            }
          }
        }
      }
    },
    {
      "data": {
        "objectId": "0xf314c9f8b3b997ad7a9454b590c71b1d50d7d67aa8c7b7a6cdc3d595b55e297f",
        "version": "1",
        "digest": "PZca4BJVr3gWLPYXLURmpNLmEHkMrDz5Td8kUZqv2Rv",
        "type": "0x2::coin::Coin<0x2::iota::IOTA>",
        "owner": {
          "AddressOwner": "0x00a059cf7241f92b9c716dd5bc2077b32a180a784d907359d007f00cb76970e1"
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "content": {
          "dataType": "moveObject",
          "type": "0x2::coin::Coin<0x2::iota::IOTA>",
          "fields": {
            "balance": "30000000000000000",
            "id": {
              "id": "0xf314c9f8b3b997ad7a9454b590c71b1d50d7d67aa8c7b7a6cdc3d595b55e297f"
            }
          }
        }
      }
    }
  ],
  "iota client objects 0x0000000000000000000000000000000000000000000000000000000000000000",
  [],
  "iota client object 0x5",
  [
    {
      "data": {
        "objectId": "0x0000000000000000000000000000000000000000000000000000000000000005",
        "version": "1",
        "digest": "Hh6D3MbK7ufnUgrABumJ2qzKNrk9KG9VyBnWo2yYjiUN",
        "type": "0x3::iota_system::IotaSystemState",
        "owner": {
          "Shared": {
            "initial_shared_version": 1
          }
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "content": {
          "dataType": "moveObject",
          "type": "0x3::iota_system::IotaSystemState",
          "fields": {
            "id": {
              "id": "0x0000000000000000000000000000000000000000000000000000000000000005"
            },
            "version": "1"
          }
        }
      }
    }
  ],
  "iota client object 0x5 --bcs",
  [
    {
      "data": {
        "objectId": "0x0000000000000000000000000000000000000000000000000000000000000005",
        "version": "1",
        "digest": "Hh6D3MbK7ufnUgrABumJ2qzKNrk9KG9VyBnWo2yYjiUN",
        "type": "0x3::iota_system::IotaSystemState",
        "owner": {
          "Shared": {
            "initial_shared_version": 1
          }
        },
        "previousTransaction": "DQa8r9XZcquYpjLpZ4DujsX3jFsy87iTLvC5AbR8Qfqq",
        "storageRebate": "0",
        "bcs": {
          "dataType": "moveObject",
          "type": "0x3::iota_system::IotaSystemState",
          "version": 1,
          "bcsBytes": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAUBAAAAAAAAAA=="
        }
      }
    }
  ],
  "iota client object 0x0000000000000000000000000000000000000000000000000000000000000000",
  [
    {
      "error": {
        "code": "notExists",
        "object_id": "0x0000000000000000000000000000000000000000000000000000000000000000"
      }
    }
  ],
  "iota client tx-block 5zibcom3dMckjyN16ygFwr5XNa9Exi1MmY3BQs984x1N",
  "ErrorObject { code: InvalidParams, message: \"Could not find the referenced transaction [TransactionDigest(5zibcom3dMckjyN16ygFwr5XNa9Exi1MmY3BQs984x1N)].\", data: None }",
  "iota client tx-block 11111111111111111111111111111111",
  "ErrorObject { code: InvalidParams, message: \"Could not find the referenced transaction [TransactionDigest(11111111111111111111111111111111)].\", data: None }"
]
